### ISSUE TITLE
Build Timeouts Investigation

### ISSUE THREAD

**Issue by markwilkie** [dotnet/dnceng#4234]
See below for Jeff's super informative email which seems to show a pattern for at least some of our timeouts that we're seeing.

Namely he's pointing out that the moment that the arm32 leg (building the same code) completed successfully, the arm64 leg hung.   (like for 45 minutes)

Please note the specific asks at the bottom which  would probably help.

Bottom line is that getting to the bottom of this (which will perhaps involved looping in AzDo) will really help our customers.

Mark
=================
From: Jeff Schwartz <jeffschw@microsoft.com>
Sent: Friday, October 4, 2024 4:22 PM
To: Mark Wilkie <mawilkie@microsoft.com>
Subject: Example Timeout Investigation
 
Today, Andy, Steve, and I were looking at a timeout in the official build that appears to be a stall in an otherwise reliable build leg.

 

Build: [Pipelines - Run 20241004.3 logs (azure.com)](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdev.azure.com%2Fdnceng%2Finternal%2F_build%2Fresults%3FbuildId%3D2552790%26view%3Dlogs%26j%3Dbf4a4c60-a3cd-5220-9ecc-692cdb8b6631%26t%3D07d07d14-eea6-5876-b173-31c4f4e9d14c&data=05%7C02%7Cmawilkie%40microsoft.com%7C338d8ed6f040436b96d408dce4cb7083%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638636809669082947%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C0%7C%7C%7C&sdata=QMBh9Mpk6n7Ni1fbEVMorOGulJYMXC7ThRa2BDS6ldc%3D&reserved=0)

 

Log for mono arm64 release: [dev.azure.com/dnceng/7ea9116e-9fac-403d-b258-b31fcf1bb293/_apis/build/builds/2552790/logs/4264](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdev.azure.com%2Fdnceng%2F7ea9116e-9fac-403d-b258-b31fcf1bb293%2F_apis%2Fbuild%2Fbuilds%2F2552790%2Flogs%2F4264&data=05%7C02%7Cmawilkie%40microsoft.com%7C338d8ed6f040436b96d408dce4cb7083%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638636809669106822%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C0%7C%7C%7C&sdata=FADb7DfstPCvqNRlcJmBD97algzp4Tu0eOoUiAWu5TI%3D&reserved=0)

2024-10-04T16:53:35.7135648Z   System.Linq.Expressions -> /__w/1/s/artifacts/bin/System.Linq.Expressions/Release/net10.0/System.Linq.Expressions.dll

2024-10-04T17:38:15.6283890Z ##[error]The Operation will be canceled. The next steps may not contain expected logs.

2024-10-04T17:38:15.6331797Z context canceled

 

Notice the 45 minutes stall with no recorded activity.  We then started to brainstorm all the things we would need in the logs to track it down (see below). 

But then… for comparisons we I looked into the mono arm32 release leg (same build) to see what a successful build looked like:

 

Log for mono arm32 release: [dev.azure.com/dnceng/7ea9116e-9fac-403d-b258-b31fcf1bb293/_apis/build/builds/2552790/logs/1745](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdev.azure.com%2Fdnceng%2F7ea9116e-9fac-403d-b258-b31fcf1bb293%2F_apis%2Fbuild%2Fbuilds%2F2552790%2Flogs%2F1745&data=05%7C02%7Cmawilkie%40microsoft.com%7C338d8ed6f040436b96d408dce4cb7083%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638636809669121929%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C0%7C%7C%7C&sdata=0LY7PJoy0%2FhKOS579DbEl6OcssuE6mxrAD9Xw2Y3MH4%3D&reserved=0)

2024-10-04T16:53:53.7748860Z Build succeeded.

2024-10-04T16:53:53.7749114Z     0 Warning(s)

2024-10-04T16:53:53.7749813Z     0 Error(s)

2024-10-04T16:53:53.7750120Z

2024-10-04T16:53:53.7750450Z Time Elapsed 00:11:14.55

2024-10-04T16:53:54.1050167Z

2024-10-04T16:53:54.1159633Z ##[section]Finishing: Build product (linux_arm)

 

Ok, now notice a coincidence in the timestamps.  The moment that the arm32 leg (building the same code) completed successfully, the arm64 leg hung. Wha????

 

Surely a coincidence, but then hours later another build with the same pattern failed: [Pipelines - Run 20241004.3 logs (azure.com)](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdev.azure.com%2Fdnceng%2Finternal%2F_build%2Fresults%3FbuildId%3D2552790%26view%3Dlogs%26j%3Dbf4a4c60-a3cd-5220-9ecc-692cdb8b6631%26t%3D07d07d14-eea6-5876-b173-31c4f4e9d14c&data=05%7C02%7Cmawilkie%40microsoft.com%7C338d8ed6f040436b96d408dce4cb7083%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638636809669135962%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C0%7C%7C%7C&sdata=f9FGdX4BeXmPLwA%2FDvNETl1BHi1BRlgtxX4DIGB6zow%3D&reserved=0)

 

And I bet, it would not take me long to go find several more timeouts this week with the same pattern.  As far as we know, there is no cross build communication that would cause a leg/msbuild to pause in this way. 

I feel we need some AzDo foo or the AzDo team to investigate to understand the cause of this timeout.

 

Thoughts?

 

Jeff

 

 

Things we know:

The last component to build was at 16:53 and then at 17:38 the build timed out.
This is the Arm64 mono leg, whereas the Arm mono leg built fine right before.
 

Things we do not know:

What is actually building at the time of the timeout.
What was running at the time of the timeout.
 

Things I think it would be nice to know:

What was actively being built?  (we would need to investigate if that is possible, and doing so without flooding the build logs)
What was running at the time of the timeout? (azdo?)
Would be nice to have a process tree dump with: process name, command line, start time, etc.
A dump? (azdo?)
This is actually hard for a few reasons:
I know there has been resistance in the past for dumps on build infra
Of all the processes running, which one?  (perhaps we could have a short allow list: csc, clang, cl, etc.)
What was the timeout set too?  (azdo?)

### Release Note Category
- [ ] Feature changes/additions 
- [ ] Bug fixes
- [ ] Internal Infrastructure Improvements
### Release Note Description

**Comment by markwilkie** [comment]
cc/ @steveisok @jeffschwMSFT @agocke @ilyas1974

**Comment by jeffschwMSFT** [comment]
apologies, the above has a typo - can you use the following link as the second example?

https://dev.azure.com/dnceng/internal/_build/results?buildId=2551267&view=results

The following legs have the same pattern:
linux-arm64 release Mono_LLVMAOT LLVMAOT
linux-x64 release Mono_LLVMAOT LLVMAOT

**Comment by dougbu** [comment]
ICM: [Incident 552521004](https://portal.microsofticm.com/imp/v5/incidents/details/552521004/summary) acknowledged

**Comment by dougbu** [comment]
@jeffschwMSFT the ICM was resolved back in November and they suggested reopening it when you have debug logs available. if you don't have such logs, should we close this tracking issue❓

**Comment by jeffschwMSFT** [comment]
I agree, we can close.  We tried to catch it with debug logs, but being non-deterministic we never did.  With the increased capacity, the frequency has gone significantly down.
